python/lsst/ts/m2gui/display/figure_constant.py

```diff
@@ -123,45 +147,250 @@ def _set_default_series(self, range_x_min, range_x_max, num_points):
+            value_min, value_max = self._get_range_points(idx)
+            values.append(value_min)
+            values.append(value_max)
+
+        self._value_y_min = min(values)
+        self._value_y_max = max(values)
+        self._update_range_axis_y(threshold=1)
+
+    def _get_range_points(self, idx):
+        """Get the range of points in a specific series.
+
+        Parameters
+        ----------
+        idx : `int`
+            Index of the series.
+
+        Returns
+        -------
+        `float`
+            Minimum value.
+        `float`
+            Maximum value.
+        """
+
+        points = self.get_points(idx)
+
+        values = list()
+        for point in points:
+            values.append(point.y())
+
+        return min(values), max(values)
+
+    def append_data(self, value, idx=0):
+        """Append the data. This is only useful if the realtime data is
+        applied.
+
+        Parameters
+        ----------
+        value : `float`
+            New coming data.
+        idx : `int`, optional
+            Index of the series. (the default is 0)
 
-        series = self.chart().series()[0]
-        count = series.count()
-        if num_x > count:
-            raise ValueError(f"Input x number(={num_x}) > available points (={count}).")
+        Raises
+        ------
+        `RuntimeError`
+            If not for the realtime data.
+        """
+
+        if not self._is_realtime:
+            raise RuntimeError("Can only be called for the realtime data.")
 
-        points = [QPointF(x, y) for (x, y) in zip(list_x, list_y)]
+        points = self.get_points(idx)
+        self._append_point(points, value)
+
+        # Replace the points in series
+        series = self.get_series(idx)
         series.replace(points)
 
-        self.axis_y.setRange(min(list_y) - self.OFFSET_Y, max(list_y) + self.OFFSET_Y)
+        # Update the range of axes to show the new data
+        if self.axis_x.max() < points[-1].x():
+            self.axis_x.setRange(points[0].x(), points[-1].x())
+
+        # Check the range of y-axis needs to be updated or not
+        self._value_y_max = max(self._value_y_max, value)
+        self._value_y_min = min(self._value_y_min, value)
+        self._update_range_axis_y()
+
+    def _append_point(self, points, value):
+        """Append the point.
+
+        Parameters
+        ----------
+        points : `list [PySide2.QtCore.QPointF]`
+            Points.
+        value : `float`
+            New coming data.
+        """
+
+        length_points = len(points)
+
+        # No data yet, begin from 0
+        if length_points == 0:
+            points.append(QPointF(0, value))
+
+        # The holded amount of data is less than the allowed maximum. Add the
+        # new point.
+        elif length_points < self._num_points:
+            points.append(QPointF(points[-1].x() + 1, value))
+
+        # Instead of instantiating new point, pop the oldest point, update the
```

Comment:
Although that might seem as clever hack, I will rather avoid that. The complexity is actually better if you don't reuse:

```
In [9]: def a(x,y): 
   ...:     p = s.points().pop() 
   ...:     p.setX(x) 
   ...:     p.setY(y) 
   ...:     s.points().append(p) 
   ...:                                                                                                                               

In [36]: def b(x,y): 
    ...:     points = s.points() 
    ...:     points.pop() 
    ...:     points.append(QPointF(x, y)) 
    ...:     s.replace(points) 
    ...:      
    ...:      
    ...:                                                                                                                              

In [37]: %timeit b(10,20)                                                                                                             
55.8 µs ± 618 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [38]: %timeit a(10,20)                                                                                                             
57.7 µs ± 404 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
_a is pop/setX/setY/append, b is pop/append(QPointF)..feel free to try yourself_

not to mention that discarding unused object is always better then reusing it, as you might hit unwanted side effects.

Reply:
It looks like the difference is under the tolerance of computer differences. Under this case, it is easier and simpler to use a new object directly.

Fixed in the commit: [a142e4c](https://github.com/lsst-ts/ts_m2gui/pull/14/commits/a142e4c44edda4f9ceebea4ed4a28648c7289a94). Thanks!